man/meteopoll.1
.TH METEOPOLL "1" "December 2001" "Meteo station tools" Othello
.SH NAME
meteopoll \- retrieve meteo data from a station
.SH SYNOPSIS
meteopoll [
.B \-d
] [
.B \-V
] [
.B \-F
] [
.BI \-l logurl
] [
.BI \-p pidfile
] [
.BI \-f conffile
] [
.BI \-m mapfile
] [
.BI \-x xmloutlet
]
.BI \-s stationname 
.SH DESCRIPTION
The 
.BR meteopoll (1)
programm connects to the weather station specified with the 
.B -s
option and the configuration file
.I conffile
and starts reading data from it. It computes averages and extreme values
for all the parameters it finds over a period of one minute, and constructs
an SQL update query to write the information into the database.
Depending on the database backend preferences set with the
.B \-b
option, the query is either sent directly to the database (as specified
in the configuration file), or to a message queue (unix domain socket
for systems that don't have System V IPC), where it will later be picked
up by the
.BR meteodequeue (1)
program, or appended to the file specified in the
.I <updatefile>
tag in the configuration file.

The 
.B \-x
option allows to specify additional outlets for the data gathered by
.BR metepoll (1).
The argument to this option is either a file or a pipe command, 
XML-formatted data is written to the file or piped into the command
every minute.
.SH OPTIONS
.TP
.B \-d
increase the debug level by one. Debug messages are written to standard
error.
.TP
.BI \-b dest
set preference for database backend communication. The 
.B \-b
option may appear multiple times. When meteopoll connects to the database
backend, it tries the different methods in the order of the corresponding
.B \-b
options. It is usually best to use a message queue as the backend,
and not to the database directly. Since the
.I stationdata
table may be locked during a long running query, 
.BR meteopoll (1)
may loose data from the weather station. The use of the message queue
in conjunction with the
.BR meteodequeue (1)
program thus improves reliabilty and performance by decreasing contention
on the 
.I stationdata
table.
Possible values for this option are
.B msgqueue
for System V message queues (or Unix domain sockets on systems without
System V IPC),
.B mysql
to write directly to the MySQL database,
.B file
to write queries to a file that can be later fed into the database and 
.B debug
to simple send the queries to the debug log.
.TP
.BI \-f conffile
Use 
.I conffile
to read the configuration, see 
.BR meteo.xml (5)
for the configuration syntax.
.TP
.BI \-m mapfile
map all data read from the station in XML form to the
.IR mapfile .
.TP
.B \-F
stay in foreground. Useful during debugging.
.TP
.BI \-l logurl
send log messages to the destination described by the 
.IR logurl .
Log URLs are of the form 
.I file:///path/to/logfile
for logging to a file, or
.I syslog:facility
to send log records to syslog facility
.IR facility .
See 
.BR meteo (1)
for details.
.TP
.BI \-p pidfile
When running in daemon mode, write the process id of the 
.BR meteopoll (1)
process to this file. Defaults to
.IR /var/run/meteopoll-<station>.pid .
.TP
.BI \-s stationname
Name of the station from which to read data. This argument is mandatory.
.TP
.B \-V
print version information and exit.
.TP
.BI \-x xmloutlet
Add 
.I xmloutlet
as an outlet specification.
If the string is a simple file name, 
.BR meteopoll (1)
writes the data not only to the database but also to this
file in XML format.
If the outlet specification starts with a
pipe symbol, the rest of the string is interpreted as a command
which is run whenever 
.BR meteopoll (1)
has new data (roughly every minute). Such a command may not use
much time, if it does not return within about a second, the next
data point from the station may be lost. Therefore such outlet
commands should never perform any activities that could block
on network latencies or lock files.
There may be arbitrarily many outlets, but keep in mind that
the time it takes to process all outlets should not exceed the
data interval from the station.

.SH SIGNALS
.BR meteopoll (1)
handles the following signals in a nontrivial way: 
.TP
SIGINT and SIGTERM
These signal not only kill the parent process, but also the currently
running child.
.TP
SIGHUP
This signal causes the parent to send a termination signal to the child.
The parent then starts a new child. Note that this signal does not
cause 
.BR metepoll (1)
to reread the configuration file.

.SH FILES
.BR meteopoll (1)
writes its process id by default to a file 
.BI /var/run/meteopoll- station .pid

.SH NOTES
.BR metepoll (1)
replaces the program
.BR meteoloop (1)
in earlier versions of the meteo packet.

.SH "SEE ALSO"
.BR meteo (1)
.BR meteo.xml (5),
.BR meteodraw (1),
.BR meteodequeue (1),
.BR meteojob (1),
.BR meteoavg (1).

.SH AUTHOR
Dr. Andreas Mueller <afm@othello.ch>


NEWS
0.9.17	a new outlet interface can be used to process meteo station data
	directly. In contrast to the map interface, the data is pushed
	to a file or into a pipe by the meteopoll process every minute.

0.9.15	new interface which allows to read current readings from the console
	to display on the web page. The new -m option to meteopoll makes
	sure current data is written to a file (the argument to the option)
	in XML format. Since updates to this file are very frequent, one
	should read the file atomically, one should use the meteomap program
	to read the data.
	fix of a nasty bug in the week label code, causing navigation failure
	at week resolution in some cases

0.9.14	gallery files included as samples
	can now compute sunrise and sunset (through meteosunriset and Sun.cc)

0.9.13	minor bug in meteobrowser.php
	gallery pages

0.9.12	some minor fixes, plus a security fix: remote code execution in
	meteobrowser.php because of missing escapeshellcmd

0.9.11	some bug fixes for GroWeather

0.9.10	turn of CRC0 for GroWeather and possibly other type E weatherlinks.

0.9.9	A hack to protect from IE reading the complete meteobrowser.php
	space offline

0.9.8	fixes a integer overflow bug: to compute the time labels for the
	maps, GraphWindow.cc used the formular (start + end)/2, but since
	about the middle of january, this causes an overflow, so that the
	target ends up in february 1936. Replace that by start/2 + end/2.
	
0.9.7	some bug fixes
	removal of offset configuration variable
	added averages configuration in XML file, so the user now gets full
	control about which averages are compute, and how
	battery and transmitter fields now work
	sunrise and sunset can also be stored in the database (this is a
		waste of space, but who cares about a few gibabytes these
		days)
	added channel of type sun, which indicates the time during which
		the sun is above the horizon

0.9.6	some bug fixes
	removal of header table

0.9.5	some minor fixes

(for further news please check the ChangeLog)

0.9.0	Major rewrite of many aspects of meteo sponsored by the University
	Univates in Lajeado in Rio Grande do Sul, Southern Brazil. Further
	contributions: Jeff Wabik (new derived value computations).

--
$Id: NEWS,v 1.32 2009/01/10 19:00:23 afm Exp $


src/meteopoll.cc
/*
 * meteopoll.cc -- successor to meteoloop, polls the station using the loop
 *                 command, and writes data to mysql database. Is aware of
 *                 nonmetric units
 *
 * (c) 2003 Dr. Andreas Mueller, Beratung und Entwicklung
 *
 * $Id: meteopoll.cc,v 1.29 2008/09/07 15:18:52 afm Exp $
 */
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif
#ifdef HAVE_STDIO_H
#include <stdio.h>
#endif
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_SYS_WAIT_H
#include <sys/wait.h>
#endif
#include <Station.h>
#include <StationFactory.h>
#include <Configuration.h>
#include <MeteoException.h>
#include <Mapfile.h>
#include <XmlOutletFactory.h>
#include <QueryOutlet.h>
#include <Datasink.h>
#include <Daemon.h>
#include <mdebug.h>
#include <errno.h>
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#include <fstream>
#include <iostream>
#include <printver.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_SIGNAL_H
#include <signal.h>
#endif
#include <sys/resource.h>

#define	ALARMTIMEOUT	30

pid_t	clpid;
bool	allowfork = true;;

// signal handler to kill child when parent is sent a signal
void	kill_child(int cause) {
	switch (cause) {
	case SIGTERM:
	case SIGINT:
		if (allowfork)
			kill(clpid, cause);
		mdebug(LOG_INFO, MDEBUG_LOG, 0, "caught signal, exiting");
		exit(EXIT_SUCCESS);
		break;
	case SIGHUP:
		mdebug(LOG_INFO, MDEBUG_LOG, 0,
			"caught SIGHUP, killing child");
		if (allowfork)
			kill(clpid, SIGTERM);
		break;
	}
	return;
}

static void	loop(const std::string& station,
	const std::string& mapfilename, const meteo::stringlist& xmloutlets) {
	// limit memory to 16MB so this process cannot monopolize the system
#define	memmax	0x1 << 25
	struct rlimit	l;
	getrlimit(RLIMIT_STACK, &l);	// stack segment
	if (l.rlim_max > memmax)
		l.rlim_max = memmax;
	setrlimit(RLIMIT_STACK, &l);

	getrlimit(RLIMIT_DATA, &l);	// data segment
	if (l.rlim_max > memmax)
		l.rlim_max = memmax;
	setrlimit(RLIMIT_DATA, &l);

	// create a new station object (this always returns something
	// connected)
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "starting new station '%s'",
		station.c_str());
	meteo::Station	*s = meteo::StationFactory().newStation(station);
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "station %s ready", station.c_str());

	// add the mapfile if mapfilename is set
	if (!mapfilename.empty()) {
		meteo::Mapfile	*mapfile = new meteo::Mapfile(mapfilename, true);
		mapfile->setStationname(station);
		s->addMapfile(mapfile);	// takes ownership
	}

	// if we have an XML file, create a XmlOutlet with a delegate
	for (meteo::stringlist::const_iterator i = xmloutlets.begin(); i != xmloutlets.end(); i++) {
		if (!i->empty()) {
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "preparing XML outlet to %s",
				i->c_str());
			meteo::XmlOutlet *xmloutlet = meteo::XmlOutletFactory::get(station, *i);
			s->addOutlet(xmloutlet); // takes ownership
		}
	}

	// create a data sink (new implementation uses QueryOutlet class)
	s->addOutlet(new meteo::QueryOutlet(station));

	// loop:
	meteo::Timeval	looptime; looptime.now();
	int	minute = looptime.getMinute();
	while (true) {
		// start the loop, expecting 10 data packets
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "request 10 packets");
		s->startLoop(10);

		while (s->expectMorePackets()) {
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
				"wait for more packets");
			// reset the alarm timer: if reading the packet takes
			// longer than 10 seconds, the connection is probably
			// broken
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "resetting timer");
			alarm(ALARMTIMEOUT);

			// read a packet from the station
			std::string	p = s->getPacket();
			mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
				"received packet of length %d", p.length());

			// update the data from the packet in the station
			s->update(p);

			// if the minute has increased by one since the last
			// update save the current record in the database and
			// reset the datarecord
			looptime.now();
			if (minute != looptime.getMinute()) {
				// send data to outlets
				s->sendOutlets(looptime.getTimekey());

				// reset and restart
				s->reset();
				minute = looptime.getMinute();
			}
		}
	}
}

static void	usage(void) {
	printf(
"usage: meteopoll [ -dFVh? ] [ -l logurl ] [ -f conffile ] [ -b prefs ] \\\n"
"       [ -p pidfile ] [ -m mapfile ] [ -x xmlfile ] -s stationname\n"
"   -d                increase debug level\n"
"   -F                don't fork (used for debugging)\n"
"   -V                print version and exit\n"
"   -h, -?            print this help screen and exit\n"
"   -f conffile       use conffile, see meteo.xml(5)\n"
"   -s stationname    let this process connect to station named stationname\n"
"   -b prefs          add backend preference prefs (one of msgqueue, mysql\n"
"                     file, debug)\n"
"   -p pidfile        write the process pid to this file\n"
"   -m mapfile        keep current sensor information in mapfile\n"
"   -x xmlfile        send output to XML file as well\n");
}

static int	meteopoll(int argc, char *argv[]) {
	std::string	conffile(METEOCONFFILE);
	std::string	logurl("file:///-");	// logging to stderr
	std::string	station, mapfilename;
	meteo::stringlist	xmloutlets;
	meteo::stringlist	preferences;
	std::string	pidfileprefix;

	// parse command line
	int	c;
	while (EOF != (c = getopt(argc, argv, "dl:f:m:s:b:p:VFh?x:")))
		switch (c) {
		case 'd':
			debug++;
			break;
		case 'l':
			logurl = std::string(optarg);
			break;
		case 'f':
			conffile = std::string(optarg);
			break;
		case 's':
			station = std::string(optarg);
			break;
		case 'b':
			preferences.push_back(std::string(optarg));
			break;
		case 'p':
			pidfileprefix = std::string(optarg);
			break;
		case 'V':
			fprintver(stdout);
			exit(EXIT_SUCCESS);
			break;
		case 'F':
			allowfork = false;
			break;
		case 'h':
		case '?':
			usage();
			exit(EXIT_SUCCESS);
		case 'm':
			mapfilename = std::string(optarg);
			break;
		case 'x':
			xmloutlets.push_back(std::string(optarg));
			break;
		default:
			mdebug(LOG_ERR, MDEBUG_LOG, 0,
				"option %c not implemented", c);
			break;
		}

	// set up logging
	mdebug_setup("meteopoll", logurl.c_str());
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "set up logging to %s",
		logurl.c_str());

	// it is an error if station name or conf
	if (station.empty()) {
		fprintf(stderr, "%s: station name not set\n", argv[0]);
		mdebug(LOG_CRIT, MDEBUG_LOG, 0, "station name not set");
		exit(EXIT_FAILURE);
	}

	// pid file
	if (pidfileprefix.empty())
		pidfileprefix = "/var/run/meteopoll-";

	// fork if necessary
	meteo::Daemon	d(pidfileprefix, station, !allowfork);

	// open the configuration object
	meteo::Configuration	conf(conffile);
	mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "read configuration %s",
		conffile.c_str());

	// create the Datasink object based on the configuration and prefs
	if (0 == preferences.size()) {
		preferences.push_back("debug");
	}
	meteo::DatasinkFactory	dsf(preferences);

	// install signal handlers
	signal(SIGTERM, kill_child);
	signal(SIGINT, kill_child);
	signal(SIGHUP, kill_child);

	// loop: fork a process and wait for it to die, exit if the child
	// exits naturally
	int	status;
	bool	giveup = false;
	do {
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "start loop");

		// fork only if allowed to do so
		if (allowfork)
			clpid = fork();
		else
			clpid = 0;

		// if the fork is not possible, we have some serious problem
		if (clpid < 0) {
			mdebug(LOG_CRIT, MDEBUG_LOG, MDEBUG_ERRNO,
				"fork failed: %s", strerror(errno));
			exit(EXIT_FAILURE);
		}

		// the child simply starts the loop
		if (clpid == 0) {
			if (allowfork)
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"starting new child");
			// reset signal dispositions so that child can easily
			// be killed
			signal(SIGHUP, SIG_DFL);
			signal(SIGTERM, SIG_DFL);
			signal(SIGINT, SIG_DFL);

			// start looping
			try {
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"starting loop(%s)", station.c_str());
				loop(station, mapfilename, xmloutlets);
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"return from loop");
			} catch (meteo::MeteoException& e) {
				mdebug(LOG_ERR, MDEBUG_LOG, 0,
					"MeteoException(%s, %s)", 
					e.getReason().c_str(),
					e.getAddinfo().c_str());
				// when we exit at this point, the parent
				// process will fork a new child (if there is a
				// a parent, that is), and it will know that we
				// had a problem
				exit(EXIT_FAILURE);
			}
			// exiting at this point means that for some reason loop
			// ended without error. This should not really happen,
			// but to tell the parent that something different happend
			// we exit with EXIT_SUCCESS
			exit(EXIT_SUCCESS);
		} // end of child code

		// here we wait for the child to terminate
		if (waitpid(clpid, &status, 0) < 0) {
			mdebug(LOG_ERR, MDEBUG_LOG, 0, "error during wait: %s",
				strerror(errno));
		} else {
			// here we should be a bit more verbose, so that it
			// becomes less difficult to find out what happened
			// to the child
			if (WIFEXITED(status)) {
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"child %d exited with status %d",
					clpid, WEXITSTATUS(status));
				giveup = (WEXITSTATUS(status) == 0);
			}
			if (WIFSIGNALED(status)) {
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"child %d killed by signal %d",
					clpid, WTERMSIG(status));
			}
			if (WIFSTOPPED(status)) {
				// this should definitely not happen, lets kill
				// the process and start a new one
				mdebug(LOG_DEBUG, MDEBUG_LOG, 0,
					"child %d stopped by signal %d",
					clpid, WSTOPSIG(status));
				kill(clpid, SIGKILL);
			}
			mdebug(LOG_CRIT, MDEBUG_LOG, 0,
				"child process died: %d", WEXITSTATUS(status));
		}
		// if there is a problem with the station, we should wait a
		// few seconds so the station can recover from the problem.
		// This also prevents meteopoll from looping. 
		mdebug(LOG_DEBUG, MDEBUG_LOG, 0, "sleeping for 3 seconds to "
			"give station time to recover from error condition");
		sleep(3);
	} while (!giveup);

	// child exited ok, so we do the same
	exit(EXIT_SUCCESS);
}

// main(argc, argv)	the main function is only needed as a wrapper to
//			catch an Exception and print a nice error message
//			on standard error (nothing is logged in this case
//			as it is unlikely that logging has been initialized
//			already
int	main(int argc, char *argv[]) {
	try {
		meteopoll(argc, argv);
	} catch (meteo::MeteoException& me) {
		fprintf(stderr, "MeteoException in meteopoll: %s/%s\n",
			me.getReason().c_str(), me.getAddinfo().c_str());
		exit(EXIT_FAILURE);
	}
	exit(EXIT_SUCCESS);
}
